File: 1471.py
import sys

sys.setrecursionlimit(100000)

N = int(input())
visited = [False] * (N + 1)
dp = [0] * (N + 1)
stack = []


def nextVertex(n):
    SUM = n
    while n:
        SUM += n % 10
        n //= 10
    if SUM > N:
        SUM -= N
    return SUM


def dfs(v):
    if dp[v]:
        return dp[v]
    if visited[v]:
        cycle = []
        cycleSize = 0
        while 1:
            cycleSize += 1
            cycle.append(stack.pop())
            if cycle[-1] == v:
                break
        for c in cycle:
            dp[c] = cycleSize
        return dp[v]
    visited[v] = True
    stack.append(v)
    temp = dfs(nextVertex(v))
    if dp[v] == 0:
        dp[v] = temp + 1
    return dp[v]


result = 0
for i in range(1, N + 1):
    if visited[i]:
        continue
    result = max(result, dfs(i))
print(result)

File: 1471.cpp
#include <bits/stdc++.h>
#define FastIO ios::sync_with_stdio(false), cin.tie(NULL), cout.tie(NULL)
using namespace std;

bool visited[200001];
int dp[200001];
vector<int> vec;
int N;

int nextVertex(int n){
    int sum = n;
    while(n != 0){
        sum += n % 10;
        n = n / 10;
    }
    if(sum > N) sum -= N;
    return sum;
}

int dfs(int v){
    if(dp[v]){
        return dp[v];
    }
    else if(visited[v]){
        vector<int> cycle;
        int cycleSize = 0;
        while(true){
            cycleSize += 1;
            cycle.push_back(vec.back());
            vec.pop_back();
            if(cycle.back() == v) break;
        }
        for(int i: cycle){
            dp[i] = cycleSize;
        }
        return dp[v];
    }
    visited[v] = true;
    vec.push_back(v);
    int temp = dfs(nextVertex(v));
    if(dp[v] == 0) dp[v] = temp + 1;
    return dp[v];
}

int main() {
    FastIO;
    int result = 0;
    cin >> N;
    for(int i = 1; i <= N; i++){
        if(visited[i]) continue;
        vec.clear();
        dfs(i);
        result = result > dp[i] ? result : dp[i];
    }
    cout << result;
}
